### PR TITLE
Implement uploads process [ch25130]

### PR DESCRIPTION
**Pull request by macharmi**

### CODE REVIEW

File: generic.go [hunk 1]
```diff
@@ -42,6 +47,86 @@ func (api API) create(path string, input interface{}, output interface{}) error
 	return wrapErrors(res, body, errs)
 }
 
+func (api API) prepareMultiPartRequest(path string, filePath string, input interface{}) (*http.Request, error) {
```
Comment (viktorchukhantsev) [inline generic.go:50]: Let's split it on several functions. File manipulations we can move out or sure.

File: generic.go [hunk 2]
```diff
@@ -42,6 +47,86 @@ func (api API) create(path string, input interface{}, output interface{}) error
...
+}
+
+// UPLOAD
+func (api API) upload(path string, filePath string, input interface{}, output interface{}) error {
+	var body []byte
+	var errs []error
+	var res *http.Response
+
+	// nolint:errcheck
+	backoff.Retry(func() error {
```
Comment (viktorchukhantsev) [inline generic.go:107]: Can we limit retries amount? I don't want that queue stuck, better reschedule upload for later.

**Review by viktorchukhantsev — CHANGES_REQUESTED** [review]

File: csv_upload.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+const (
+	uploadEndoint = "data_sources/:data_source_uuid/uploads"
+)
+
+// UploadCSVFile Upload a CSV file to data platform
+func (api API) UploadCSVFile(filePath string, uploadRequest *CsvUploadRequest) (*CsvUploadResponse, error) {
+	result := &CsvUploadResponse{}
+
+	path := strings.Replace(uploadEndoint, ":data_source_uuid", uploadRequest.DataSourceUUID, 1)
```
[the remark this answers is not in the document]
Comment (macharmi) [reply csv_upload.go:32]: `strings.Replace()` is used to modify paths everywhere. In term of clarity for the reader `:data_source_uuid` is always better than `%v`

File: generic.go (hunk 2 above)
Comment (macharmi) [reply generic.go:107]: Currently, it stops after 15 minutes.

```
/usr/local/go/bin/go tool test2json -t /tmp/___upload_file_test_go -test.v -test.run ^\QTestUploadCSVFile\E$
=== RUN   TestUploadCSVFile
2020/09/16 14:06:06 2020-09-16 14:06:06.121223553 +0200 CEST m=+0.002179253
2020/09/16 14:06:06 2020-09-16 14:06:06.674076786 +0200 CEST m=+0.555032647
2020/09/16 14:06:07 2020-09-16 14:06:07.755027443 +0200 CEST m=+1.635983290
2020/09/16 14:06:09 2020-09-16 14:06:09.06558626 +0200 CEST m=+2.946542084
2020/09/16 14:06:10 2020-09-16 14:06:10.648333481 +0200 CEST m=+4.529289314
2020/09/16 14:06:12 2020-09-16 14:06:12.989308527 +0200 CEST m=+6.870264358
2020/09/16 14:06:17 2020-09-16 14:06:17.49599238 +0200 CEST m=+11.376948229
2020/09/16 14:06:20 2020-09-16 14:06:20.717877365 +0200 CEST m=+14.598833193
2020/09/16 14:06:26 2020-09-16 14:06:26.327007145 +0200 CEST m=+20.207962975
2020/09/16 14:06:33 2020-09-16 14:06:33.977248868 +0200 CEST m=+27.858204700
2020/09/16 14:06:49 2020-09-16 14:06:49.372462159 +0200 CEST m=+43.253418014
2020/09/16 14:07:18 2020-09-16 14:07:18.644155048 +0200 CEST m=+72.525110885
2020/09/16 14:08:15 2020-09-16 14:08:15.458574316 +0200 CEST m=+129.339530178
2020/09/16 14:08:58 2020-09-16 14:08:58.31652074 +0200 CEST m=+172.197476658
2020/09/16 14:09:51 2020-09-16 14:09:51.157508457 +0200 CEST m=+225.038464324
2020/09/16 14:10:40 2020-09-16 14:10:40.24156692 +0200 CEST m=+274.122522715
2020/09/16 14:11:38 2020-09-16 14:11:38.385065803 +0200 CEST m=+332.266021662
2020/09/16 14:12:25 2020-09-16 14:12:25.375419278 +0200 CEST m=+379.256375036
2020/09/16 14:13:12 2020-09-16 14:13:12.961872629 +0200 CEST m=+426.842828480
2020/09/16 14:14:23 2020-09-16 14:14:23.708330242 +0200 CEST m=+497.589285956
2020/09/16 14:15:06 2020-09-16 14:15:06.826452941 +0200 CEST m=+540.707408771
2020/09/16 14:15:49 2020-09-16 14:15:49.01813329 +0200 CEST m=+582.899089123
2020/09/16 14:16:40 2020-09-16 14:16:40.671687438 +0200 CEST m=+634.552643179
2020/09/16 14:17:44 2020-09-16 14:17:44.924524345 +0200 CEST m=+698.805480084
2020/09/16 14:19:06 2020-09-16 14:19:06.683125207 +0200 CEST m=+780.564081034
2020/09/16 14:19:54 2020-09-16 14:19:54.27040026 +0200 CEST m=+828.151356090
2020/09/16 14:20:42 2020-09-16 14:20:42.095703412 +0200 CEST m=+875.976659270
2020/09/16 14:21:57 2020-09-16 14:21:57.250590685 +0200 CEST m=+951.131546580
    upload_file_test.go:40: Request error: i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout; i/o timeout
--- FAIL: TestUploadCSVFile (951.14s)
``` 

But we can tweak it (see https://godoc.org/github.com/cenkalti/backoff#ExponentialBackOff)

**Comment by pkopac** [comment]
@macharmi `alpha`, not `main`

File: README.md [hunk 4]
```diff
@@ -194,6 +194,15 @@ api.MetricsListSubscriptions(&Cursor{}, "customerUUID")
 api.MetricsListActivities(&Cursor{}, "customerUUID")
 ```
 
+### CSV Upload
+
+Available CSV Upload methods:
+
+
+```go
+api.UploadCSVFile(filePath string, uploadRequest *CsvUploadRequest)
```
Comment (pkopac) [inline README.md:203]: You should probably also implement a method that will upload from a `io.Reader`. Not everyone wants to first dump the data on the disk. But there's a catch if you need to retry, you'd need to rewind.

@viktorchukhantsev how will you do it in the service?

**Review by pkopac — CHANGES_REQUESTED** [review]
I won't approve this into main, we're merging all CSV-related stuff into `alpha` ;)

**Review by pkopac — CHANGES_REQUESTED** [review]
alpha

File: README.md (hunk 4 above)
Comment (macharmi) [reply README.md]: changed it to an `interface{}` that accepts a `string` considered as a file path or an `io.Reader`

**Review by pkopac — APPROVED** [review]

**Review by viktorchukhantsev — APPROVED** [review]

**Comment by macharmi** [comment]
@pkopac @viktorchukhantsev I just added alpha to `.Travis` file.

File: a/a.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package main
```
Comment (pkopac) [inline a/a.go:1]: is this here by an accident?

**Pull request merged**